## 💻 Exercises: Module 9

### Exercises: Level 1

1. Declare a function _add_two_numbers_. It takes two parameters and it returns a sum. 

In [1]:
def two_num(a, b):
    print("The sum of two numbers is:", a + b)
    
two_num(10, 40)

The sum of two numbers is: 50


2. Area of a circle is calculated as follows: area = π x r x r. Write a function that calculates _area_of_circle_.

In [1]:
def area_of_circle(r):
    pi = 3.14
    area = pi * r**2
    return area
print(f'The Area of circle is: {area_of_circle(7)}')

The Area of circle is: 153.86


3. Write a function called add_all_nums which takes arbitrary number of arguments and sums all the arguments. Check if all the list items are number types. If not do give a reasonable feedback.

In [8]:
def add_all(*numbs):
    total = sum(numbs)
    return total
print(f'The addition of numbers: {add_all(25, 25, 29, 70)}')
print("______________________")
#OR 
def addition_all(*nums):
    total = 0
    for num in nums:
        total +=num
    return total
print(addition_all(25, 19, 40, 70, 98))


The addition of numbers: 149
______________________
252


4. Temperature in °C can be converted to °F using this formula: °F = (°C x 9/5) + 32. Write a function which converts °C to °F, _convert_celsius_to-fahrenheit_.

In [11]:
#Define a function name call CONVERT_CELSIUS_TO_FAHRENHEIT

def convert_celsius_to_fahrenheit(c):           #Pass a prementer of Celsius
    f = (c * 9 / 5) + 32                        #assigning the calculation to the block
    return f
print(convert_celsius_to_fahrenheit(50))          #Call the function to perform the task

122.0


5. Write a function called check-season, it takes a month parameter and returns the season: Autumn, Winter, Spring or Summer.

In [21]:
def check_season(month):
    '''Season of Month'''
    month = month.title()
    if month in ('September', 'October', 'November'):
        return "Autumn"
    elif month in ('December', 'January', 'Fabuary'):
        return'Winter'
    elif month in ('March', 'April', 'May'):
        return "Spring"
    elif month in ('June', 'July', 'August'):
        return "Summer"
    else:
        return "The month does not exist check spelling month"    
    
print(check_season('april'))

Spring


6. Write a function called calculate_slope which return the slope of a linear equation

In [26]:
# Formula of slope of linear equation
# slope_m = y2 - y1 / x2 - x1

def slop_m(y2, y1, x2, x1):
    slop_m = (f'The Slope of a given value is: {y2 - y1 / x2 - x1}')       #Calculating the slope
    return slop_m

# Call function
print(slop_m(25, 10, 50, 20))
print("_____________________________")

#using key value or Position argument
print(slop_m(x1=25, x2= 90, y1= 80, y2=40))

The Slope of a given value is: 4.800000000000001
_____________________________
The Slope of a given value is: 14.111111111111114


7. Quadratic equation is calculated as follows: ax² + bx + c = 0. Write a function which calculates solution set of a quadratic equation, _solve_quadratic_eqn_.

In [33]:
# using quadritic formula ax² + bx + c = 0.

#declear a function called solved_quadratic_eqn

import math
def solve_quadratic_eqn(a, b, c):
    
    # Calculate the discriminant
    discriminant = b**2 - 4*a*c
    
    # check if discriminant is non - negative numbers
    if discriminant >=0:
        root1 = (-b + math.sqrt(discriminant)) / (2 * a)
        root2 = (-b - math.sqrt(discriminant)) / (2 * a)
        return (root1, root2)
    else:
        # if discriminant is negative return complex roots
        real_part = -b / (2 * a)
        other_part = math.sqrt(-discriminant) / (2 * a)
        return (real_part + other_part, real_part - other_part)

print(f"The Roots are: {solve_quadratic_eqn(9, 6, 27)}")

The Roots are: (1.3663398378642617, -2.0330065045309285)


8. Declare a function named print_list. It takes a list as a parameter and it prints out each element of the list.

In [41]:
# declear list function

def print_lists(*items):
    for item in items:
        print(item.title())
    
print_lists('suleiman', 'abas')

Suleiman
Abas


9. Declare a function named reverse_list. It takes an array as a parameter and it returns the reverse of the array (use loops).

```py
print(reverse_list([1, 2, 3, 4, 5]))
# [5, 4, 3, 2, 1]
print(reverse_list1(["A", "B", "C"]))
# ["C", "B", "A"]
```

In [10]:
def reverse_list(*items):
    reverse_list = []
    for x in range (len(items) -1, -1, -1):
        reverse_list.append(items[x])
    return reverse_list
    
print(reverse_list(25, 4, 6, 8, 10))
print(reverse_list("A", "C", "D", "E"))

[10, 8, 6, 4, 25]
['E', 'D', 'C', 'A']


10. Declare a function named capitalize_list_items. It takes a list as a parameter and it returns a capitalized list of items

In [19]:
# 10. Declare a function named capitalize_list_items. It takes a list as a parameter and it returns a capitalized list of items

def capitalize_list_items(lst):
    capitalize_items = []
    for i in lst:
        capitalize_items.append(i.capitalize())
    return capitalize_items

print(capitalize_list_items(['abas', 'suleiman', 'lahadi', 'mufidat', 'abdulrahman']))

['Abas', 'Suleiman', 'Lahadi', 'Mufidat', 'Abdulrahman']


11. Declare a function named add_item. It takes a list and an item parameters. It returns a list with the item added at the end.

```py
food_staff = ['Potato', 'Tomato', 'Mango', 'Milk']
print(add_item(food_staff, 'Meat'))     # ['Potato', 'Tomato', 'Mango', 'Milk','Meat']
numbers = [2, 3, 7, 9]
print(add_item(numbers, 5))      [2, 3, 7, 9, 5]

In [22]:
def add_items(list, item):
    new_list = []
    for element in list:
        new_list.append(element)
    new_list.append(item)
    return new_list
food_staff = ['Potato', 'Tomato', 'Mango', 'Milk']
print(add_items(food_staff, 'Meat'))
numbers = [2, 3, 7, 9]
print(add_items(numbers, 5))

['Potato', 'Tomato', 'Mango', 'Milk', 'Meat']
[2, 3, 7, 9, 5]


2. Declare a function named remove_item. It takes a list and an item parameters. It returns a list with the item removed from it.

```py
food_staff = ['Potato', 'Tomato', 'Mango', 'Milk']
print(remove_item(food_staff, 'Mango'))
numbers = [2, 3, 7, 9]
print(remove_item(numbers, 3))  

In [29]:
def remove_items(items, ritem):
    remove = []
    for item in items:
        remove.append(item)
    remove.remove(ritem)
    return remove

food_staff = ['Potato', 'Tomato', 'Mango', 'Milk']
print(remove_items(food_staff, 'Mango'))
numbers = [2, 3, 7, 9]
print(remove_items(numbers, 3)) 

['Potato', 'Tomato', 'Milk']
[2, 7, 9]


13. Declare a function named sum_of_numbers. It takes a number parameter and it adds all the numbers in that range.

```py
print(sum_of_numbers(5))  # 15
print(sum_of_numbers(10)) # 55
print(sum_of_numbers(100)) # 5050

In [4]:
def sum_of_numbers(numbers):
    total = 0
    for num in range(numbers + 1):
        total +=num
    return total
print(sum_of_numbers(5))  # 15
print(sum_of_numbers(10)) # 55
print(sum_of_numbers(100)) # 5050

15
55
5050


14. Declare a function named sum_of_odds. It takes a number parameter and it adds all the odd numbers in that range.

In [15]:
#14. Declare a function named sum_of_odds. It takes a number parameter and it adds all the odd numbers in that range.

def sum_odd(x):
    odd = 0
    for i in range(x + 1):
        if i % 2 != 0:
            odd +=i
    return odd
print(f"The Sum of odd number in 5 is: {sum_odd(5)}")
print(f"The Sum of odd number in 10 is: {sum_odd(10)}")
print(f"The Sum of odd number in 100 is: {sum_odd(100)}")

The Sum of odd number in 5 is: 9
The Sum of odd number in 10 is: 25
The Sum of odd number in 100 is: 2500


15. Declare a function named sum_of_even. It takes a number parameter and it adds all the even numbers in that - range

In [13]:
#15. Declare a function named sum_of_even. It takes a number parameter and it adds all the even numbers in that - range.

def sum_of_even(n):
    even = 0
    for number in range(n + 1):
        if number % 2 == 0:
            even +=number
    return even
    
print(f'The sum of even number in 5 is: {sum_of_even(5)}')
print(f'The sum of even number in 10 is: {sum_of_even(10)}')
print(f'The sum of even number in 100 is: {sum_of_even(100)}')

The sum of even number in 5 is: 6
The sum of even number in 10 is: 30
The sum of even number in 100 is: 2550


In [16]:
print("______COMPLETE LEVEL ON_________")

______COMPLETE LEVEL ON_________


### Exercises: Level 2

1. Declare a function named evens_and_odds . It takes a positive integer as parameter and it counts number of evens and odds in the number.

```py
    print(evens_and_odds(100))
    # The number of odds are 50.
    # The number of evens are 51.
```

In [21]:
# 1. Declare a function named evens_and_odds . It takes a positive integer as parameter and it counts number of evens and odds in the number.

def evens_and_odds(y):
    evens = 0
    odds = 0
    for e in range(y + 1):
        if e % 2 == 0:
            evens +=1
        else:
            odds +=1
    return evens, odds

print(f'The total even and odd number in 100 are: {evens_and_odds(100)}')

The total even and odd number in 100 are: (51, 50)


2. Call your function factorial, it takes a whole number as a parameter and it return a factorial of the number

In [27]:
#Call your function factorial, it takes a whole number as a parameter and it return a factorial of the number
def factorial(f):
    if f == 1 or f == 0:
        return 1
    else:
        return f * factorial(f - 1)
print(factorial(7))

5040


3. Call your function _is_empty_, it takes a parameter and it checks if it is empty or not

In [39]:
def is_empty(text):
    if text:
        return True
    else:
        return False
print(is_empty("Abdulrahman"))
print(is_empty([]))

True
False


4. Write different functions which take lists. They should calculate_mean, calculate_median, calculate_mode, calculate_range, calculate_variance, calculate_std (standard deviation).

In [24]:
def calculate_mean(numbers):
    numbers.sort()
    sum_num = sum(numbers)
    length = len(numbers)
    mean = sum_num / length
    return mean
    
# call the function
print("The mean of calculated value are as follows:")
print(calculate_mean([25,26,40,23,45]))
print(calculate_mean([21,2,2,5,8]))
print(calculate_mean([25,22,10,3,20]))
print("_______________________________")

#Alternative for def mean function
def calculate_mean2(l):
    total = 0
    for x in l:
        total +=x
    mean = total / len(l)
    return mean    
print("The mean of calculated value are as follows: using the method 2")
print(calculate_mean2([25,26,40,23,45]))
print(calculate_mean2([21,2,2,5,8]))
print(calculate_mean2([25,22,10,3,20]))
print("_______________________________")

# define a function
def calculate_median(numbers):
    numbers.sort() # sort the numbers from smallest to largest
    length = len(numbers)
    median0 = numbers[length//2] # median if the number length is odd
    if length % 2 == 0:  # check for even if the length is even
        median1 = numbers[length//2 - 1]  # first median number
        median2 = numbers[length//2]       # second median number
        median = (median1 + median2)/2
        return median
    else:
        return numbers[length//2]

# function call
print('\nThe median for the following list of values.')
print(calculate_median([1,8,2,3,23,4,5,6,7]))



def calculate_range(numbers):
    '''Function to calculate the range of values'''
    if len(numbers) == 0: # check if list length == 0, 
        return ('Invalide range')
    else:
        max_value = max(numbers)
        min_value = min(numbers)
        range = max_value - min_value
        return range
    
    
    # function call
print(calculate_range([1,3,7,13,18])) 
# function that calculate variance
def calculate_variance(numbers):
    '''Function to calculate the variance of a list of numbers'''
    if len(numbers) == 0:  # Check if the list is empty
        return None  # Return None if the list is empty

    # Calculate the mean
    mean = sum(numbers) / len(numbers)

    # Calculate the squared differences from the mean
    squared_differences = [(x - mean) ** 2 for x in numbers]

    # Step 3: Calculate the variance
    variance = sum(squared_differences) / len(numbers)

    return variance

def calculate_standard_deviation(numbers):
    '''Function to calculate the standard deviation of a list of numbers'''
    variance = calculate_variance(numbers)

    if variance is None:  # If variance is None, return None
        return None

    # Step 1: Calculate the standard deviation
    std_deviation = variance ** 0.5  # Square root of variance

    return std_deviation

# Function calls
numbers = [2, 4, 6, 8, 10]

# Calculate variance
variance = calculate_variance(numbers)
print(f"Variance: {variance}") 

# Calculate standard deviation
std_deviation = calculate_standard_deviation(numbers)
print(f"Standard Deviation: {std_deviation:.4f}")  


The mean of calculated value are as follows:
31.8
7.6
16.0
_______________________________
The mean of calculated value are as follows: using the method 2
31.8
7.6
16.0
_______________________________

The median for the following list of values.
5
17
Variance: 8.0
Standard Deviation: 2.8284


In [25]:
def calculate_mode(numbers):
    '''Function to calculate the mode of a list of numbers'''
    if len(numbers) == 0:  # Check if the list is empty
        return None  # Return None if the list is empty

    # Create a dictionary to count occurrences of each number
    frequency = {}
    for number in numbers:
        if number in frequency:
            frequency[number] += 1  # Increment count if the number exists in the dictionary
        else:
            frequency[number] = 1  # Initialize count to 1 if the number is new

    # Find the number(s) with the highest frequency
    max_frequency = max(frequency.values())  # Find the highest count
    mode = [key for key, value in frequency.items() if value == max_frequency]  # Get all keys with the highest count

    # If multiple numbers have the same highest frequency, return all of them
    if len(mode) == 1:
        return mode[0]  # Return the single mode
    return mode  # Return a list of modes if there are multiple

# Function calls
numbers1 = [1, 2, 2, 3, 4, 4, 5]
numbers2 = [1, 1, 1, 2, 3, 3, 3]
numbers3 = []

print(f"Mode: {calculate_mode(numbers1)}")  
print(f"Mode: {calculate_mode(numbers2)}") 
print(f"Mode: {calculate_mode(numbers3)}") 

Mode: [2, 4]
Mode: [1, 3]
Mode: None


### Exercises: Level 3

1. Write a function called is_prime, which checks if a number is prime.

In [27]:
def is_prime(number):
    '''function to check for prime number'''
    if number == 2:         #  if number == 2, it prime
        return (f'The number, {number} is prime number')
    elif number < 2:
        return (f'The number, {number} is not prime number')
    for i in range(2,number +1):                    # looping througt range of values inclusive of the number to check for
        if number % i == 0:
            return (f'The number, {number} is not prime number')
        else:
            return (f'The number, {number} is prime number')
# function call
print(is_prime(40))

The number, 40 is not prime number


2. Write a functions which checks if all items are unique in the list.

In [28]:
def unique_values(numbers):
    '''Function to check for unique values. len(numbers)==len(set(numbers))'''
    empty_set = set()  # empty set to remove duplicate of numbers
    for number in numbers:     # loop through the list of numbers
        if number in empty_set:            # check for number in list of numbers
            return False  
        else:
            empty_set.add(number)
    return True
# function call
print(unique_values([2,5,4,4]))
print(unique_values([2, 5, 2, 7, 8]))  
print(unique_values([2, 5, 7, 8])) 

False
False
True


3.  Write a function which checks if all the items of the list are of the same data type.

In [29]:
def same_data_type(items):
    '''attempt to check all data types of a list'''
    if not items: # handling empty list
        return True 
    first_item = type(items[0]) # checking the data type of the first item in a list of items
    for item in items:             # loop througt list of items
        if type(item) != first_item:      # check the first and last item contain the same data type
            return False
    else:
        return True
# function call
print(same_data_type([1,2,3,4]))
print(same_data_type([1,'mango',3,4]))
print(same_data_type([]))

True
False
True


4.  Write a function which check if provided variable is a valid python variable

In [30]:
import keyword

def is_valid_variable(var_name):
    '''Function to check if a provided variable name is valid in Python'''
    if len(var_name) == 0:
        return ('No variable name provied')
    if keyword.iskeyword(var_name): 
        return False  # Return False if it's a Python reserved keyword
    if var_name.isidentifier():  
        return True  # Return True if it follow identifier rules
    else:
        return False  # If it passes all checks, it is valid

# Function calls
print(is_valid_variable("variable1"))    
print(is_valid_variable("1variable"))    
print(is_valid_variable("for"))          
print(is_valid_variable("_private_var")) 
    

True
False
False
True


5. Go to the data folder and access the countries-data.py file.

- Create a function called the most_spoken_languages in the world. It should return 10 or 20 most spoken languages in the world in descending order
- Create a function called the most_populated_countries. It should return 10 or 20 most populated countries in descending order.

In [31]:
from countries_data import countries_data
from collections import Counter

def most_spoken_languages(data, top_n=10):
    language_counts = Counter()  # A dictionary-like object to count occurrences
    
    for country in data:  # Loop through all countries
        language_counts.update(country["languages"])  # Count languages
    
    # Get the top `top_n` languages sorted by count in descending order
    return language_counts.most_common(top_n)

# Most spoken languages
print("Most Spoken Languages:")
for language, count in most_spoken_languages(countries_data, top_n=10):
    print(f"{language}: {count}")

Most Spoken Languages:
English: 91
French: 45
Arabic: 25
Spanish: 24
Portuguese: 9
Russian: 9
Dutch: 8
German: 7
Chinese: 5
Serbian: 4


In [32]:
def most_populated_countries(data, top_n=10):
    # Sort the countries by population in descending order
    sorted_countries = sorted(data, key=lambda x: x["population"], reverse=True)
    
    return sorted_countries[:top_n]  # Return the top `n` countries

# Most populated countries
print("\nMost Populated Countries:")
for country in most_populated_countries(countries_data, top_n=10):
    print(f"{country['name']}: {country['population']:,}")


Most Populated Countries:
China: 1,377,422,166
India: 1,295,210,000
United States of America: 323,947,000
Indonesia: 258,705,000
Brazil: 206,135,893
Pakistan: 194,125,062
Nigeria: 186,988,000
Bangladesh: 161,006,790
Russian Federation: 146,599,183
Japan: 126,960,000


### LIST COMPREHENSION EXERCISES

1. Filter only negative and zero in the list using list comprehension

2. Flatten the following list of lists of lists to a one dimensional list :

   ```py
   list_of_lists =[[[1, 2, 3]], [[4, 5, 6]], [[7, 8, 9]]]

   output
   [1, 2, 3, 4, 5, 6, 7, 8, 9]

Using list comprehension create the following list of tuples:
   ```py
   [(0, 1, 0, 0, 0, 0, 0),
   (1, 1, 1, 1, 1, 1, 1),
   (2, 1, 2, 4, 8, 16, 32),
   (3, 1, 3, 9, 27, 81, 243),
   (4, 1, 4, 16, 64, 256, 1024),
   (5, 1, 5, 25, 125, 625, 3125),
   (6, 1, 6, 36, 216, 1296, 7776),
   (7, 1, 7, 49, 343, 2401, 16807),
   (8, 1, 8, 64, 512, 4096, 32768),
   (9, 1, 9, 81, 729, 6561, 59049),
   (10, 1, 10, 100, 1000, 10000, 100000)]
   ```

4. Flatten the following list to a new list:
   ```py
   countries = [[('Finland', 'Helsinki')], [('Sweden', 'Stockholm')], [('Norway', 'Oslo')]]
   output:
   [['FINLAND','FIN', 'HELSINKI'], ['SWEDEN', 'SWE', 'STOCKHOLM'], ['NORWAY', 'NOR', 'OSLO']]
   ```

5. Change the following list to a list of dictionaries:
   ```py
   countries = [[('Finland', 'Helsinki')], [('Sweden', 'Stockholm')], [('Norway', 'Oslo')]]
   output:
   [{'country': 'FINLAND', 'city': 'HELSINKI'},
   {'country': 'SWEDEN', 'city': 'STOCKHOLM'},
   {'country': 'NORWAY', 'city': 'OSLO'}]
   ```

6. Change the following list of lists to a list of concatenated strings:
   ```py
   names = [[('Asabeneh', 'Yetayeh')], [('David', 'Smith')], [('Donald', 'Trump')], [('Bill', 'Gates')]]
   output
   ['Asabeneh Yetaeyeh', 'David Smith', 'Donald Trump', 'Bill Gates']
   ```

7. Write a lambda function which can solve a slope or y-intercept of linear functions.

##### Complete Module 9